Story post backend

voidxnull opened the pull request:
Part of #877.

To create a story post, set `type` to `'story'` and `text_type` to either `'markdown'` or `'html'`. Next, back-end processes `text_source` according to `text_type` and outputs the result into `text`. Keep in mind that some controllers that return a collection of posts do not include `text` and `text_source` by default anymore. Add `?fields=+text,text_source` query to include those fields (prefix with `+` to extend default list of fields).

The PR also contains a complete refactoring of `createPost` and `updatePost` controllers. The motivation for this change is that it's getting harder to manage and validate all text post invariants (short_text, long_text, story). Each type of a post requires a different treatment and it makes the `createPost` and `updatePost` controllers too convoluted.

I'm also working on validation of posts using Joi. The PR may be merged as it is, I going to make a separate PR later, but for now validation errors are being sent with the `500` code instead of `400`.

- [x] Make `title` optional.
- [x] Save first image url to `more.image.url`
- [x] Save first two paragraphs of text to `more.shortText`
- [x] Exclude `text` and `text_source` columns from post lists. (all GET `/posts/*` methods)
- [x] Columns: `text_source`, `text_type`, `text`
  - [x] For non-story posts, save text into both `text_source` and `text`
  - [x] For story posts, `text` contains processed text which is safe to be rendered using `dangerouslySetInnerHTML` 
  - [x] `text_type` can only be `markdown` or `html`. No plain text.
- [ ] Fix controllers once #868 is merged.

The client side needs to be refactored to use `more.shortText` instead of `text`.

voidxnull committed:
Add tests for createPost and updatePost controllers

Only for story posts.

voidxnull committed:
Add transaction support to Post.attach/detach/update[tag] methods

voidxnull committed:
Refactor createPost and updatePost controllers

Most of the business logic has been moved to model methods.
Queries are wrapped in a transaction.
Implemented rendering of text_source into text for story posts.

voidxnull commented:
Travis is failing because of the linting error in the stable branch. Here is a fix: https://github.com/Lokiedu/libertysoil-site/pull/822/commits/e2f4f5d9ec244f3ef1c520c3c6c0d10db1f2b26f

Can we also stop coveralls from spamming into our PRs?

xekc commented:
➤ Vladimiras Lekecinskas commented:

@voidxnull do you think a rich (dropbox paper-like or medium.com-like) editor can be used here? What needs to be done on the back end further?

xekc commented:
➤ Vladimiras Lekecinskas commented:

@indeyets could you merge this in?

voidxnull commented:
@xekc Absolutely. I think it's done on the back-end: now posts may be written either in Markdown or in HTML, and we already have an image loading API.

voidxnull commented:
Working on replacing checkit with https://github.com/hapijs/joi. The reason for this is checkit's limitations, lack of documentation, and bugs like this https://github.com/tgriesser/checkit/issues/76#issuecomment-277145908. I'm happy with the API of Joi so far.

voidxnull committed:
Fix CreatePost and EditPost components according to new API requirements

Use `text_source` instead of `text` to specify text.

voidxnull committed:
Add ability for some post controllers to specify returned fields

- Added applyFieldsQuery helper.
- Now post listing controllers don't select the `text` and
  `text_source` columns by default.
  Add `?fields=+text,text_source` query param to include omitted
  columns.

voidxnull committed:
Extract first image url into more.image.url on post save

voidxnull committed:
Fix and refactor test/integration/anonym-access.js
